### PR TITLE
Clean up HTTP async iterator code

### PR DESCRIPTION
**Pull request by ry**
cc @kevinkassimo @keroxp

### CODE REVIEW

File: util/async.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
+
+// TODO(ry) It'd be better to make Deferred a class that inherits from
+// Promise, rather than an interface. This is possible in ES2016, however
```
Comment (piscisaureus) [inline util/async.ts]: Also, this kicks v8 in a mode where every promise is very slow. Don't do it.

File: http/server.ts [hunk 2]
```diff
@@ -313,65 +289,61 @@ async function readRequest(
...
+    const bufr = new BufReader(httpConn);
+    while (true) {
+      const [req, err] = await readRequest(httpConn, bufr);
+      if (err) {
+        // TODO(ry) This should be more granular. Perhaps return back a 400 or
+        // 500 error?
+        httpConn.close();
+        break;
+      }
+      this.queue.push(req);
```
Comment (piscisaureus) [inline http/server.ts]: This does not provide proper backpressure semantics. 
You should make sure that readRequest is not getting called when the iterator isn't being iterated.
Basically, the 'queue' thing needs to go away.

File: util/async.ts (hunk 1 above)
Comment (ry) [reply util/async.ts]: I'm just moving this code to a centralized location. I'll leave it to future work to optimize it.

File: http/server.ts (hunk 2 above)
Comment (ry) [reply http/server.ts]: Done, using your latch code. PTAL.

**Comment by kevinkassimo** [comment]
@ry seems with this implementation, there is always going to be either 0 or just 1 elements in sendQueue/recvQueue?

**Comment by ry** [comment]
@kevinkassimo I think so too, but I haven't been able to come up with a simplification.

**Comment by piscisaureus** [comment]
I'm sorry but this still creates an unbounded queue.

Imagine multiple physical TCP connections coming in while no requests are being processed.
They're all going to call .send() on the latch which makes that queue grow without backpressure.

**Comment by piscisaureus** [comment]
I'll write something myself

**Comment by piscisaureus** [comment]
See https://github.com/denoland/deno_std/pull/411/commits/29f3e8e7cb7b4cd535554d2a37ceb57ca7bb6320
Can still be cleaned up a lot further. A particular issue is that there's no "clean" way for a keepalive tcp connection to close.
Just showing what the general pattern should be.

File: http/server.ts [hunk 3]
```diff
@@ -1,43 +1,21 @@
...
-  }
-  return {
-    promise,
-    resolve,
-    reject
-  };
-}
+import { Channel, deferred, Deferred } from "../util/async.ts";
 
 interface HttpConn extends Conn {
```
Comment (piscisaureus) [inline http/server.ts]: This should be a class, not an interface.

File: http/server.ts [hunk 4]
```diff
@@ -264,114 +248,83 @@ export class ServerRequest {
...
  * See https://v8.dev/blog/fast-async
  */
-async function readRequest(
-  c: HttpConn,
-  bufr?: BufReader
-): Promise<[ServerRequest, BufState]> {
-  if (!bufr) {
-    bufr = new BufReader(c);
-  }
+async function* iterateHttpRequests(
```
Comment (piscisaureus) [inline http/server.ts]: This should be a method on HttpConn

File: http/server.ts [hunk 5]
```diff
@@ -264,114 +248,83 @@ export class ServerRequest {
...
-    env.reqQueue = [];
-    for (const result of queueToProcess) {
-      yield result;
-      // Continue read more from conn when user is done with the current req
-      // Moving this here makes it easier to manage
-      serveConn(env, result.conn, result.r);
+
+      for await (const [req, err] of iterateHttpRequests(httpConn)) {
+        if (err) {
+          // TODO(ry) This should be more granular. Perhaps return back a 400 or
```
Comment (piscisaureus) [inline http/server.ts]: Preferrably `iterateHttpRequests` would just yield ServerRequest objects. Handling errors on the connection should be the responsibility of the `iterateHttpRequests()` function or the `HttpConn` class.

File: http/server.ts [hunk 6]
```diff
@@ -249,102 +249,82 @@ export class ServerRequest {
...
+    let [s, err]: [string, BufState] = await tp.readLine();
+    if (err) {
+      yield [null, err];
+      return;
+    }
 
-  return [req, err];
+    [req.method, req.url, req.proto] = s.split(" ", 3);
+    [req.headers, err] = await tp.readMIMEHeader();
+    yield [req, err];
```
Comment (piscisaureus) [inline http/server.ts]: Todo: let's say the request has a body which is processed by the consumer of this iterator.
We'd want to wait for this processing to be complete before reading a new set of headers.
Therefore we might need an await statement after this `yield`, e.g. `await req.done` where `done` is a Promise that's resolved when the request has been processed.

File: http/server.ts [hunk 7]
```diff
@@ -264,114 +248,83 @@ export class ServerRequest {
...
+    // Set a new pipeline deferred associated with this request
+    // for future requests to wait for.
+    c.pendingDeferredMap.set(req.pipelineId, deferred());
+
+    req.conn = c;
+    req.r = bufr!;
+    req.w = bufw;
+
+    // First line: GET /index.html HTTP/1.0
+    const tp = new TextProtoReader(bufr!);
```
Comment (piscisaureus) [inline http/server.ts]: TextProtoReader construction could probably be moved to the top of the generator.

**Comment by piscisaureus** [comment]
I realize there is one big issue with this, which is that one tcp connection can block processing of others.

**Comment by piscisaureus** [comment]
... which has been fixed now.

File: http/server.ts (hunk 7 above)
Comment (ry) [reply http/server.ts]: No, it should be constructed per header. I've moved it back.

File: http/server.ts (hunk 6 above)
Comment (ry) [reply http/server.ts]: Added the TODO to the source.

File: http/server.ts (hunk 3 above)
Comment (ry) [reply http/server.ts]: I've added this TODO to the source code.

File: http/server.ts (hunk 5 above)
Comment (ry) [reply http/server.ts]: Maybe. The server is modeled after the Go server, which returns errors explicitly. Although not very javascript-y, this made it easy to port code over. I wouldn't be opposed to someone going through and changing these returned errors into exceptions - but I don't think it hurts how it currently is organized.

**Comment by ry** [comment]
@piscisaureus I like the MuxAsyncIterator abstraction. I've moved it to `util/async.ts` and did a few other clean ups. Please take another look.

**Comment by piscisaureus** [comment]
@ry wdyt about https://github.com/denoland/deno_std/pull/411/commits/eab25c54f488b743906c62178773a81b67b3ea1b ?

**Comment by ry** [comment]
Looks fine. I’m slightly worried it doesn’t work when you respond to a request out of order? But I think we have tests for that...

**Comment by piscisaureus** [comment]
> Looks fine. I’m slightly worried it doesn’t work when you respond to a request out of order? But I think we have tests for that...

That's no longer possible, because no new requests will be accepted on the same connection before the previous one has done responding.
This is also how http works in practice. The server could e.g. respond with 'connection: close' and it wouldn't even be valid for the client to send another request on the same connections, so clients always have to wait for a response before sending a new request.

**Comment by zekth** [comment]
Is this PR may fix the issue of malformed or encrypted mime headers?
https://github.com/denoland/deno_std/pull/411/files#diff-2e4c2adf52a4573569ea4cebb26b523eR235

Returning an error 400 if the request can't be properly read.

**Comment by piscisaureus** [comment]
@zekth I don't know. Is there a test for this issue?

**Comment by zekth** [comment]
> @zekth I don't know. Is there a test for this issue?

Not atm because we use `Headers` in test and not raw HTTP which don't give use the flexibility of sending anything we want in the request.

But i can write some like done in :  https://github.com/denoland/deno_std/blob/master/http/racing_server_test.ts

For ref: https://github.com/denoland/deno/issues/2346

**Comment by kevinkassimo** [comment]
This PR results in a 50% reduction of req/sec:

master:
```
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   353.40us  153.27us   2.88ms   95.73%
    Req/Sec    14.48k   692.05    15.67k    68.32%
  290989 requests in 10.10s, 13.88MB read
Requests/sec:  28809.59
Transfer/sec:      1.37MB
```

this branch:
```
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    24.05ms   61.05ms 353.56ms   88.58%
    Req/Sec    11.01k     4.44k   15.41k    81.50%
  195783 requests in 10.05s, 9.34MB read
Requests/sec:  19489.73
Transfer/sec:      0.93MB
```
Latency seems exceptionally problematic.
I believe this is somehow related to we removed pipelining code? Not so sure if `wrk` does pipelining...

**Comment by piscisaureus** [comment]
@kevinkassimo pipelining hasn't been removed but I'll look into it.

**Comment by piscisaureus** [comment]
@kevinkassimo Can you try again? I think the latest patch should remove the unfair scheduling that was introduced by the MuxAsyncIterator and restore overall performance.

However I'm still seeing pretty bad max-latency values. Note that it's strictly the *maximum* latency that is quite bad. The overall distribution (the one you get with `wrk http://localhost:4500 --latency`) is actually looking quite good. So there's some request somewhere that's taking a bit longer. I suspect it may be that this new code might effectively accept 10 connnections  first before servicing any requests on them, which makes that the first request takes a bit of a hit. But I'm not sure.

**Comment by piscisaureus** [comment]
It could also be due to v8 doing some optimizing after the benchmark has started. I'm seeing much better max-latency numbers on the second and subsequent runs.

**Comment by kevinkassimo** [comment]
@piscisaureus Seems performing better than master now.

This branch:
```
./third_party/wrk/mac/wrk -d 10 http://localhost:4500
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   285.13us  185.29us   4.83ms   94.48%
    Req/Sec    18.73k   799.01    20.61k    74.75%
  376607 requests in 10.10s, 17.96MB read
Requests/sec:  37284.09
Transfer/sec:      1.78MB
```

**Comment by kevinkassimo** [comment]
There is another small issue I noticed: it seems under 500 concurrent connections our HTTP server would die silently... Might be unrelated to this PR though:

```
$ ./third_party/wrk/mac/wrk -c 500 -d 10 http://localhost:4500
Running 10s test @ http://localhost:4500
  2 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.34ms   28.00ms 535.90ms   98.24%
    Req/Sec    17.87k     2.44k   22.63k    74.00%
  355802 requests in 10.01s, 16.97MB read
  Socket errors: connect 0, read 446, write 13, timeout 0
Requests/sec:  35534.95
Transfer/sec:      1.69MB
$ ./third_party/wrk/mac/wrk -c 500 -d 10 http://localhost:4500
Running 10s test @ http://localhost:4500
  2 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     0.00us    0.00us   0.00us     nan%
    Req/Sec     0.00      0.00     0.00       nan%
  0 requests in 10.10s, 0.00B read
  Socket errors: connect 0, read 485, write 14, timeout 0
Requests/sec:      0.00
Transfer/sec:       0.00B
```

(Actually on master this would result in a panic on
```
thread '<unnamed>' panicked at 'bad rid', ../../cli/resources.rs:237:15
```
)

**Comment by piscisaureus** [comment]
@ry I think async test may not actually be running

**Comment by piscisaureus** [comment]
@kevinkassimo I think we may not be dealing with EMFILE errors properly. What happens if you increase the ulimit on your system?

**Comment by piscisaureus** [comment]
Or running out of ephemeral ports.

**Comment by ry** [comment]
I will land this branch now and update the submodule in master.

**Pull request merged**